sklearn/linear_model/_cd_fast.pyx

```diff
@@ -712,42 +706,58 @@ def enet_coordinate_descent_multi_task(floating[::1, :] W, floating l1_reg,
                     continue
 
                 # w_ii = W[:, ii] # Store previous value
-                _copy(n_tasks, W_ptr + ii * n_tasks, 1, wii_ptr, 1)
+                _copy(n_tasks, &W[0, ii], 1, &w_ii[0], 1)
 
                 # if np.sum(w_ii ** 2) != 0.0:  # can do better
-                if _nrm2(n_tasks, wii_ptr, 1) != 0.0:
+                if (w_ii[0] != 0.):  # faster than testing full norm for non-zeros
+                    # Using Numpy:
                     # R += np.dot(X[:, ii][:, None], w_ii[None, :]) # rank 1 update
-                    _ger(RowMajor, n_samples, n_tasks, 1.0,
-                         X_ptr + ii * n_samples, 1,
-                         wii_ptr, 1, &R[0, 0], n_tasks)
+                    # Using Blas Level2:
+                    # _ger(RowMajor, n_samples, n_tasks, 1.0,
+                    #      &X[0, ii], 1,
+                    #      &w_ii[0], 1, &R[0, 0], n_tasks)
+                    # Using Blas Level1 and for loop for avoid slower threads
+                    # for such small vectors
+                    for jj in range(n_tasks):
+                        _axpy(n_samples, w_ii[jj], &X[0, ii], 1, &R[0, jj], 1)
 
+                # Using numpy:
                 # tmp = np.dot(X[:, ii][None, :], R).ravel()
-                _gemv(RowMajor, Trans, n_samples, n_tasks, 1.0, &R[0, 0],
-                      n_tasks, X_ptr + ii * n_samples, 1, 0.0, &tmp[0], 1)
+                # Using BLAS Level 2:
+                # _gemv(RowMajor, Trans, n_samples, n_tasks, 1.0, &R[0, 0],
+                #       n_tasks, &X[0, ii], 1, 0.0, &tmp[0], 1)
+                # Using BLAS Level 1 (faster small vectors like here):
+                for jj in range(n_tasks):
+                    tmp[jj] = _dot(n_samples, &X[0, ii], 1, &R[0, jj], 1)
 
                 # nn = sqrt(np.sum(tmp ** 2))
                 nn = _nrm2(n_tasks, &tmp[0], 1)
 
                 # W[:, ii] = tmp * fmax(1. - l1_reg / nn, 0) / (norm_cols_X[ii] + l2_reg)
-                _copy(n_tasks, &tmp[0], 1, W_ptr + ii * n_tasks, 1)
+                _copy(n_tasks, &tmp[0], 1, &W[0, ii], 1)
                 _scal(n_tasks, fmax(1. - l1_reg / nn, 0) / (norm_cols_X[ii] + l2_reg),
-                      W_ptr + ii * n_tasks, 1)
+                      &W[0, ii], 1)
 
                 # if np.sum(W[:, ii] ** 2) != 0.0:  # can do better
-                if _nrm2(n_tasks, W_ptr + ii * n_tasks, 1) != 0.0:
+                if (W[0, ii] != 0.):  # faster than testing full col norm
```

Comment:
It's not equivalent.
- before: equivalent to test if all the W[:, ii] == 0. If so, no need to compute the dot. Makes sense
- after: only check if W[0, ii] == 0.
I think you miss some computations here.

Reply:
you're right. It's something we haven't cleaned up.

Reply:
it's actually on purpose as it's unlikely that you have w_ii[0]==0 and some values != in w_ii[1:] so it should do as well and avoid computations

Reply:
What really needs to be avoided is the computation of the norm of W[:, ii]. I think a safe and not too heavy way to do the check is just a loop over W[:, ii], (as in https://github.com/mathurinm/celer/blob/master/celer/multitask_fast.pyx#L360)

Reply:
> it's unlikely

Unlikely is not always ? If it can happen that only `w_ii[0]==0`, the computation would be wrong. I like @mathurinm's idea.

Reply:
implemented